### PR TITLE
NMS-10191: The KSC Dashlet for the Ops-Board is not working

### PR DESCRIPTION
**Pull request by patrick-schweizer**
* JIRA: https://issues.opennms.org/browse/NMS-10191

Our [continuous integration system](http://bamboo.internal.opennms.com:8085) will test and verify your changes.

### CODE REVIEW

**Comment by patrick-schweizer** [comment]
This pull request incorporates the comments from https://github.com/OpenNMS/opennms/pull/2022
@christianpape, @mvrueden,  @j-white please check again. If ok, this PR can be merged. I will make a separate one for foundation-2018.

File: features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java [hunk 1]
```diff
@@ -425,6 +433,23 @@ public Component getComponent() {
         });
     }
 
+    String decodeResourceId(String resourceId){
+        // This is a quite ugly hack but it will be removed as of foundation-2018 when the class ResourceId will be
```
Comment (j-white) [inline features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java]: Yeah, this URL encoding caused a number of other issues too. Glad to see the ResourceId class help mitigate this.

**Comment by j-white** [comment]
Thanks for the update. Looks good to me.

File: features/vaadin-dashlets/dashlet-ksc/src/test/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashletTest.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2013-2017 The OpenNMS Group, Inc.
```
Comment (mvrueden) [inline features/vaadin-dashlets/dashlet-ksc/src/test/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashletTest.java]: Should be 2018-2018.

File: features/vaadin-dashlets/dashlet-ksc/src/test/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashletTest.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ * OpenNMS(R) is Copyright (C) 1999-2017 The OpenNMS Group, Inc.
```
Comment (mvrueden) [inline features/vaadin-dashlets/dashlet-ksc/src/test/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashletTest.java]: Should be `1999-2018`

File: features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java [hunk 4]
```diff
@@ -425,6 +433,23 @@ public Component getComponent() {
...
+            return URLDecoder.decode(resourceId, "UTF-8");
+        } catch (UnsupportedEncodingException e) {
+            // should not happen for UTF-8 but if let's throw an unchecked Exception
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    OnmsResource determineResourceByResourceId(String resourceId){
+        OnmsResource resource = m_resourceDao.getResourceById(resourceId);
+        resource =(resource.getParent()== null) ? resource : resource.getParent();
```
Comment (mvrueden) [inline features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java]: Can a resource only have one parent, or do we need to traverse the tree upwards, until no parent is found?

**Comment by RangerRick** [comment]
@mvrueden hah, I had gone to this page before you added your comments, didn't see them until I merged  :D

File: features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java (hunk 4 above)
Comment (RangerRick) [reply features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java]: Good question about the resources.  I would imagine it's technically possible to have more than one level, but I don't know that we actually do.  Easy enough to fix outside of the context of these changes.

**Pull request merged**